src/components/TaskProjectSelector.vue

```diff
@@ -0,0 +1,283 @@
+  components: {
+    NSelect,
+    NButton,
+    NIcon,
+    TrashCan,
+  },
+  props: {
+    taskId: {
+      type: String,
+      required: true,
+    },
+    memberships: {
+      type: Object as PropType<Membership[]>,
+      required: true,
+    },
+    membershipEdits: {
+      type: Object as PropType<MembershipEdits>,
+      required: true,
+    },
+  },
+  setup(props, { emit }) {
+    const asanaStore = useAsanaStore();
+    const projects = computed<Project[]>(() => asanaStore.projects);
+    const sections = ref<Section[]>();
+    const sectionsLoading = ref(false);
+    const isTaskProjectSelectorShown = ref(false);
+    const selectedProject = ref<string>();
+    const selectedSection = ref<string>();
+    const taskMemberships = ref(props.memberships);
+    const taskMembershipEdits = ref(props.membershipEdits);
+
+    const getSwimlane = (s: string) => {
+      const arraySplit = s.split(":");
+      return arraySplit.length > 1 ? arraySplit[0] : "No Swimlane";
+    };
+
+    const getSection = (s: string) => {
+      const arraySplit = s.split(":");
+      return arraySplit[1] ? arraySplit[1].split("|")[0] : s;
+    };
+
+    const openProject = (gid: string) => {
+      asanaStore.LOAD_SELECTED_PROJECT(gid);
+    };
+
+    const makeProjectOptions = (projects: Project[]) => {
+      return projects.map((p) => {
+        return {
+          label: p.name,
+          value: p.gid,
+        };
+      });
+    };
+
+    const makeSectionOptions = (sections: Section[]) => {
+      return sections.map((p) => {
+        return {
+          label: getSwimlane(p.name) + " | " + getSection(p.name),
+          value: p.gid,
+        };
+      });
+    };
+
+    const updateSections = async (val) => {
+      selectedSection.value = undefined;
+      sectionsLoading.value = true;
+      sections.value = (await getSectionsByProject(
+        val
+      )) as unknown as Section[];
+      sectionsLoading.value = false;
+    };
+
+    const getSectionsByProject = async (proj) =>
+      await asanaClient?.sections.findByProject(proj);
+
+    const addMembership = (
+      taskId: string,
+      projectId: string,
+      sectionId: string
+    ) => {
+      const project = projects.value.find((p) => p.gid === projectId);
+      const section = sections.value?.find((s) => s.gid === sectionId);
+      if (project && section) {
+        taskMemberships.value.push({
+          project: project as unknown as Resource,
+          section: section,
+        });
+      }
+
+      // create a unique edit ID based on the task and project
+      taskMembershipEdits.value[taskId + projectId] = {
+        taskId: taskId,
+        projectId: projectId,
+        sectionId: sectionId,
+        isDelete: false,
+        isDone: false,
+      };
+
+      isTaskProjectSelectorShown.value = false;
+      selectedProject.value = undefined;
+      selectedSection.value = undefined;
+
+      emit("update:memberships", taskMemberships.value);
```

Comment:
I'm not quite sure I understand why there are two events being emitted with logically similar data. Could you elaborate?

Reply:
update:memberships is used to make the task editor UI reactive to the changes. This is the final state of the task's memberships if the user saves. 
However, this array does not contain deletes (as deletion will remove the membership from this array). Storing the deletes are necessary since the Asana API only exposes add/delete functionality. There isn't a "replaceAll" endpoint that we can use.
Hence, we use membershipEdits to store the edits made by the user. This array will be utilized to make the API calls.